File: nnops/cpu/ops/unary_ops.cpp
#include <nnops/cpu/ops/unary_ops.h>
#include <nnops/tensor.h>
#include <nnops/data_type.h>
#include <nnops/common.h>

using nnops::Tensor, nnops::TensorMeta;

namespace nnops::cpu::ops {

Tensor linear(const Tensor &input, const Tensor &weight, const std::optional<Tensor> &bias) {
    NNOPS_CHECK(
        (input.shape(-1) == weight.shape(1)) && (!bias.has_value() || bias.value().shape(0) == weight.shape(0)),
        "linear input and weight(or bias) are incompatible.");
    return Tensor();
}

} // namespace nnops::cpu::ops

File: src/cpu/ops/unary_ops.rs
use crate::tensor::Tensor;
use crate::cpu::ops::matmul::matmul;
use crate::cpu::ops::binary_ops::add;

pub fn linear(input: &Tensor, weight: &Tensor, bias: Option<&Tensor>) -> Tensor {
    nnops_check!(
        input.shape_at(-1) == weight.shape_at(1),
        "linear input and weight are incompatible."
    );
    if let Some(b) = bias {
        nnops_check!(
            b.shape_at(0) == weight.shape_at(0),
            "linear bias and weight are incompatible."
        );
    }
    let w_t = weight.transpose(-1, -2);
    let mut ret = matmul(input, &w_t);
    if let Some(b) = bias {
        ret = add(&ret, b);
    }
    ret
}